Bundling JS with ESBuild

aq1018 opened the pull request:
## Issue

In [Better Lighthouse Scores](https://github.com/hugo-toha/toha/discussions/697) Discussion, I agreed to bundle JS and CSS. This PR accomplishes the JS part. A follow up PR will be created to address CSS bundling.

@hossainemruz, I had the goal of changing as little as possible while accomplishing the conversion, however, due to es6 module not allowing to write to globals, I had to refactor quite a bit of code. So, sorry if there are a lot of changes.

## Testing

I created this PR incrementally.

* First, I layed the foundations for the new paradim.
* Then I migrated each JS file or feature separately in each commit.

As I work on each aspect or feature, I tested and made sure that feature worked.

But given the amount of changes, I won't be surprised if there are bugs / regressions. 

If you find anything funny, just tag me on this PR, and I'll make sure to get it fixed.

Also, the `config.yml` file format has changed a bit. Specifically, I added a lot of switches to enable / disable certain features to allow treeshaking of unused js code. Please see comments in [layouts/partials/helpers/get-esbuild-options.html](https://github.com/hugo-toha/toha/pull/702/files#diff-4374db830881775ddaa002c14924e252e56b984fc21017dc9f142c671fb935ff) file for a more detailed explanation.

Example `config.yml`

```yaml
baseURL: https://hugo-toha.github.io

languageCode: en-us
title: "John's Blog"
theme: "toha"

# Manage languages
# For any more details, you can check the official documentation: https://gohugo.io/content-management/multilingual/
languages:
  en:
    languageName: English
    weight: 1
  bn:
    languageName: বাংলা
    weight: 3

# Force a locale to be use, really useful to develop the application ! Should be commented in production, the "weight" should rocks.
# DefaultContentLanguage: bn

# Allow raw html in markdown file
markup:
  goldmark:
    renderer:
      unsafe: true
  tableOfContents:
    startLevel: 2
    endLevel: 6
    ordered: false

# At least HTML and JSON are required for the main HTML content and
# client-side JavaScript search
outputs:
  home:
    - HTML
    - RSS
    - JSON

# Enable global emoji support
enableEmoji: true

# Site parameters
params:
  # Background image of the landing page
  background: /images/site/background.jpg

  # Provide logos for your site. The inverted logo will be used in the initial
  # transparent navbar and the main logo will be used in the non-transparent navbar.
  logo:
    main: /images/site/main-logo.png
    inverted: /images/site/inverted-logo.png
    favicon: /images/site/favicon.png

  # GitHub repo of your site
  gitRepo: https://github.com/hugo-toha/hugo-toha.github.io
  # Default branch of your Git repo
  gitBranch: main

  # Configure the number of section title visible in the top navbar
  topNavbar:
    maxVisibleSections: 5

  # Configure various features of this theme
  features:

    # Enable dark theme
    darkMode:
      enable: false
      services:
        darkreader:
          defaultColorScheme: system # options are 'system', 'dark', 'light'
          fixes:
            invert: ['img[src$=".svg"]'] # inverts svg colors.
          theme:
            brightness: 100
            contrast: 100
            sepia: 0

    # Enable and configure portfolio
    portfolio:
      enable: false

    # Enable and configure blog posts
    blog:
      enable: true
      # Share post on different social media
      shareButtons:
        facebook: true
        twitter: true
        linkedin: true
        reddit: true
        whatsapp: true
        email: true
        # tumblr: true
        # pocket: true
        # diaspora: true
        # mastodon: true

    # Enable & configure "Notes" features
    notes:
      enable: true

    # Enable comment feature. There, should be only one of them.
    comment:
      enable: false
      services:
        disqus:
          shortName: toha-example-site
        # valine:
        #   appId: HTV6askWxVo3vOxlqjjaq2hd-MsYXbMMI
        #   appKey: pVrT3C85KGIuk27t07eh6bUC
        #   avatar: retro
        #   placeholder: Share your thought.
        #   lang: en
        #   recordIP: true
        #   enableQQ: true
        # utteranc:
        #   repo: github.com/hugo-toha/hugo-toha.github.io
        #   issueTerm: title
        #   theme: github-light
        # giscus:
        #   repo: your-repo/name
        #   repoID: your-repo-id
        #   category: your-category
        #   categoryID: your-category-id
          # theme: light
          # map: url
          # reaction: 1
          # metadata: 0
          # inputPosition: bottom
          # crossOrigin: anonymous
        # commento:
        #   serverURL: cdn.commento.io
        #   autoInit: true
        #   hideDeleted: false

    # Enable Analytics
    analytics:
      enabled: false
      services:
        # Google Analytics
        google:
          id: G-H4LBG7NDFZ
        # # CounterDev
        # counterDev:
        #   id: <your counterdev id>
        # # GoatCounter
        # goatCounter:
        #   code: <your goat counter code>
        # # Matomo / Piwik
        # matomo:
        #   instance: matomo.example.com
        #   siteId: 1 # Or any other number

    # Enable Support
    support:
      enable: false
      services:
        kofi:
          user: hossainemruz
          text: Tip Me
          textColor: '#f9fafc'
          backgroundColor: '#248aaa'
    
    # specify whether you want to show Table of Contents in reading page
    toc:
      enable: true
    
    # Show tags under the post title
    tags:
      enable: true
    
    # Specify whether to show flag in the language selector. Default is true.
    flags:
      enable: true
      # # If you want to use different country flag for a language, specify them here.
      # flagOverwrites:
      #   - languageCode: en
      #     countryCode: us

    # Enable this to use `embed-pdf` shortcode.
    embedpdf:
      enable: false

    # Enable this to create flowcharts using shortcodes.
    flowchart:
      enable: false
      services:
        # Uncomment for `mermaid` shortcode.
        mermaid:
          # For config options, see: https://mermaid-js.github.io/mermaid/#/Setup?id=configuration
          # theme: dark

    # Enable this to create mathematic expressions using `$$` blocks
    math:
      enable: false
      services:
        # https://katex.org/docs/autorender.html#api for more configurations
        katex:
          delimiters:
            - left: $$
              right: $$
              display: true
            - left: \\[
              right: \\]
              display: true
            - left: $
              right: $
              display: false
            - left: \\(
              right: \\)
              display: false

    # Enable to use custom syntax highlight
    # Please note, Hugo comes with it's own html based syntax highlighter.
    # Your code block will still be syntax highlighted by hugo.
    # For more details: https://gohugo.io/content-management/syntax-highlighting/
    # To disable Hugo's builtin syntax highlight,
    # see: https://gohugo.io/getting-started/configuration-markup#highlight
    # ```
    # # config.yaml
    # markup:
    #   # this disables hugo's syntax highlighting.
    #   codeFences: false
    # ```
    syntaxHighlight:
      enable: true
      services:
        hljs:
          # see: https://highlightjs.readthedocs.io/en/latest/api.html#configure
          noHighlightRe: /^no-highlight$/i

    # Enable to use `video-player` shortcode
    videoPlayer:
      enable: false
      services:
        # convert .js-player class into video player via https://plyr.io/
        # Can play HTML5 Video, Audio, Youtube, Vimeo.
        # For more info on setup: https://github.com/sampotts/plyr#quick-setup
        plyr:
          # options doc: https://github.com/sampotts/plyr#options
          # fullscreen: true


  # Provide footer configuration.
  footer:
    enable: true
    # You can provide your custom footer template using this option.
    # Put your template in "layouts/partials" folder of your repo.
    template: footer.html
    # Show/hide navigation in the footer. Default is "true".
    navigation:
      enable: true
      # show custom menus in the footer
      customMenus: true
    # Show/hide "Contact Me" section in the footer. Default is "true".
    contactMe:
      enable: true
    # Show/hide credentials section in the footer. Default is "true".
    credentials:
      enable: true

    # Show/hide newsletter section in the footer. Default is "true".
    # Currently, it supports "mailchimp".
    newsletter:
      enable: true
      provider: mailchimp
      mailchimpURL: https://github.us1.list-manage.com/subscribe/post?u=19de52a4603135aae97163fd8&amp;id=094a24c76e

    # Show/hide disclaimer notice in the footer. Default is "false".
    disclaimer:
      enable: true

```

hossainemruz commented:
Hi @aq1018, Thank you for your hard work in this PR. Look like lot of changes. I might take some time to finish reviewing. Feel free to work on CSS PR based on this branch. Once, this PR get merged, we can rebase the CSS PR against main branch.

aq1018 commented:
That's the plan! 😄

aq1018 commented:
Hi @hossainemruz I realized that I made a few mistakes on achievements and some other feature toggle related conditionals, I added two more small commits to fix those issues.

P.s. I added  an example`config.yml` to reflect updated settings. When testing, you should use that one as a basis.

hossainemruz commented:
Okay. This is major change. It may break backward compatibility for existing users. I think we should bump a major version with this change. Feel free to make your changes. Do whatever best for the theme.

aq1018 commented:
Found video-player shortcode wasn't working. fixed.

aq1018 commented:
@hossainemruz p.s. I updated the `config.yaml` file in top comment. I missed a few flags.

aq1018 commented:
@hossainemruz This PR is ready for review. As well as the CSS bundle #705 
Let me know if you need me to change anything. Thanks!

hossainemruz commented:
Thank you @aq1018. I have released a new version of the theme with latest changes. This paves the way to merge these PRs. I have started reviewing. It may take few days though.

hossainemruz commented:
Hi @aq1018 Can you give me permission to push in this PR?

hossainemruz commented on assets/scripts/features/darkmode/index.js:
```diff
@@ -0,0 +1,60 @@
+    return await import('./darkreader');
+  }
+
+  throw Error(' No service defined for feature markMode.');
```
Possible typo. Change `markMode` to `darkMode`.

hossainemruz commented on assets/scripts/pages/note.js:
```diff
@@ -0,0 +1,32 @@
+import $ from 'jquery';
+import imagesLoaded from 'imagesloaded';
+
+$(document).ready(function () { 
```
`ready()` function seems to be deprecated. Can we move to the alternative?

hossainemruz commented on assets/scripts/pages/single.js:
```diff
@@ -0,0 +1,58 @@
+  // add scroll to top button
+  var btn = $('#scroll-to-top');
+
+  $(window).scroll(function () {
```
`scroll()` function has been deprecated. Can we move to alternative?

hossainemruz commented on assets/scripts/sections/navbar.js:
```diff
@@ -37,27 +37,25 @@ const updateNavBar = () => {
-        $('.navbar-collapse').collapse('hide');
-      });
-    }
+$(document).ready(function () {
```
`ready()` is deprecated. Let's move on alternative.

hossainemruz commented on assets/scripts/sections/navbar.js:
```diff
@@ -37,27 +37,25 @@ const updateNavBar = () => {
+  // ==================================================
+  // When the user scrolls down 80px from the top of the document, resize the navbar's padding and the logo's font size
+  // $.onscroll = function() {scrollFunction()};
+  $(document).scroll(function () {
```
`scroll()` deprecated.

hossainemruz reviewed (changes requested):
Hi @aq1018, My apologies for this late review. I was on deadline for another project. Then, there was FIFA world cup :P.  Now, I am able to spend some time with this project. Let's get it merge.

Regarding the PR. Thank you for this immense amount of work. The project is now better organized and will be easy to work with.

The PR looks good to me. There are few comments. Mostly related to use of deprecated function. Can you please take a look.

aq1018 commented:
@hossainemruz thanks for the review. I'll work on it later today.

aq1018 commented:
@ hossainemruz sorry for the long wait. I ended up doing the following:

1. installed eslint and configured it to use the `standard` + `prettier` + `no-jquery` config.
2. use that to find errors in code ( had a few missing `const`s, and some minor issues )
3. migrated away from jquery with the `eslint-config-no-jquery`'s help.
4. formatted code with eslint to have a consistent js code style.
5. rebased to latest main.

I hope you don't mind me adding the [standard](https://standardjs.com/index.html#why-should-i-use-javascript-standard-style) style guide. If you don't like `standard`, there is also other popular style guides such as `airbnb`, `google` or other standards. For more details please see this article: https://betterprogramming.pub/comparing-the-top-three-style-guides-and-setting-them-up-with-eslint-98ea0d2fc5b7

Lastly, I still haven't had the chance to test it against the example site yet. I will do more testing tomorrow.

hossainemruz reviewed (approved):
The PR looks good to me. Thank you again for this immense amount of work. 💘
